### ISSUE TITLE
Improvement suggestion: Accidental assignment instead of == when using .filter() is not reported as compile error

### ISSUE THREAD

**Issue by AntoninBrettsnajdrONSEMI** [microsoft/TypeScript#36529]
**TypeScript Version:**  3.7.5

**Search Terms:**
Accidental assignment using filter, boolean lambda, assignment lambda 

**Code**

```ts
[1,2,3].filter(num => num = 2)
```

**Expected behavior:**
Compile error, since what I wanted was: `num == 2`

**Actual behavior:**
No compile error.

**Playground Link:** 
https://www.typescriptlang.org/play/?ssl=1&ssc=31&pln=1&pc=1#code/NoRgNATGDMC6B0AzAlgGwC4FMBOAKAdgK4C2ABALwB8pRZ5pEAlEA

**Notes**
I am mainly a Kotlin/Java guy, so I am not sure if `num = 2` is a boolean expression in JS, but anyway it feels wrong to me that this was allowed by TS, since it caused an unintentional silent bug in my code which took me some time to figure out.

**Comment by AlCalzone** [comment]
> I am not sure if `num = 2` is a boolean expression in JS

It is truthy, which is why the expression is valid:
![grafik](https://user-images.githubusercontent.com/17641229/73439704-57efcf00-4350-11ea-9820-4598abe2d9eb.png)

Not sure if it is possible to make the callback argument readonly, but this would probably break a lot of code.

**Comment by AntoninBrettsnajdrONSEMI** [comment]
Ah I see ... btw I've also tried this now:
![image](https://user-images.githubusercontent.com/38497416/73440245-3b07cb80-4351-11ea-8d43-7149ad957e4f.png)
And it says "number". So that is IMO a good clue for the compiler since only booleans should be allowed here. 

When I type this in the TS playground:
```ts
var y
let x: boolean = (y=1)
```
I get the compile error which is correct. Therefore if e.g. .filter() is expecting a function returning boolean, it should cause the compile error too, shouldn't it?

**Comment by AntoninBrettsnajdrONSEMI** [comment]
Hah, but look at this: https://www.typescriptlang.org/play/#code/NoRgNABATJDMC6A6AZgSwDYBcCmAnAFAHYCuAthALwB8ERZl0AlIwLABQ7AbgIa4QCe7dNkwQAHgC5ajSjQBGAewXDuhBvhnVa-BiFZthogF5S6pKSVJy8m+UpVqKtS7edkKeoA

If I try it with lambda, it is a compile error:
```ts
var y
let x: () => boolean = () => (y = 1)
let z: (num: number) => boolean = (num) => (num=1)
```

but inside the filter(), no error reported ... strange 🤔

**Comment by AlCalzone** [comment]
`filter()` does **not** expect the predicate function to return a boolean. An element is kept in the resulting array if the predicate returns a [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) value for the element.

**Comment by AntoninBrettsnajdrONSEMI** [comment]
> `filter()` does **not** expect the predicate function to return a boolean. An element is kept in the resulting array if the predicate returns a [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) value for the element.

Ah, ok, now I get it ☺️ . I was used from Kotlin/Java that it is a boolean, so this kind of error/typo can never happen there, but in the JS world it seems a bit more complicated I guess ☺️ . If it could be addressed somehow it would be cool, but currently I have no idea how. Maybe switching from truthy value to boolean, but I am pretty sure that it would break something.

**Comment by WhileTrueEndWhile** [comment]
#36472

**Comment by jcalz** [comment]
Also relevant: #18497 and [eslint's `no-param-reassign` rule](https://eslint.org/docs/rules/no-param-reassign)

**Comment by RyanCavanaugh** [comment]
If you're new to JS it's unlikely you're going to find novel or unexplored issues in TS - please try Stack Overflow first. Thanks!
